device/core/src/client_diagnostic.ts

```diff
@@ -0,0 +1,206 @@
+    this.correlationContext = {
+      creationTimeUtc
+    };
+  }
+
+  /**
+   * @method            module:azure-iot-device.Diagnostics.getEncodedCorrelationContext
+   * @description       get the encoded diagnostic correlation context.
+   *
+   * @returns {string}
+   */
+  public getEncodedCorrelationContext(): string {
+    /* Codes_SRS_NODE_DEVICE_DIAGNOSTICS_01_003: [The getEncodedCorrelationContext function returned the encoded string of correlation context.] */
+    let encodedCorrelationContext = '';
+    for (let key in this.correlationContext) {
+      if (encodedCorrelationContext !== '') {
+        encodedCorrelationContext += Diagnostics.DIAGNOSTIC_CONTEXT_SYMBOL_AND;
+      }
+      encodedCorrelationContext += (Diagnostics.DIAGNOSTIC_CONTEXT_PROPERTY[key] + Diagnostics.DIAGNOSTIC_CONTEXT_SYMBOL_EQUAL + this.correlationContext[key]);
+    }
+    return encodedCorrelationContext;
+  }
+}
+
+/**
+ * Handling device client diagnostic information.
+ */
+export class DiagnosticClient {
+  // The base of diagnostic id character. 0-9a-zA-Z is 62.
+  static DIAGNOSTIC_ID_CHARACTER_BASE: number = 62;
+
+  static DIAGNOSTIC_TWIN_KEY_SAMPLING_PERCENTAGE: string = '_diag_sample_rate';
+  static DIAGNOSTIC_TWIN_KEY_ERROR_MSG: string = '_diag_info';
+
+  private diagSamplingPercentage: number;
+  private currentMessageNumber: number;
+
+  /**
+   * @constructor
+   */
+  constructor() {
+    /* Codes_SRS_NODE_DEVICE_DIAGNOSTICCLIENT_01_001: [The Client constructor shall initial the percentage and message count.] */
+    this.diagSamplingPercentage = 0;
+    this.currentMessageNumber = 0;
+    this.onDesiredTwinUpdate = this.onDesiredTwinUpdate.bind(this);
+  }
+
+  /**
+   * @method            module:azure-iot-device.DiagnosticClient.getDiagSamplingPercentage
+   * @description       get the diagnostic sampling percentage.
+   *
+   * @returns {number}
+   */
+  getDiagSamplingPercentage(): number {
+    /* Codes_SRS_NODE_DEVICE_DIAGNOSTICCLIENT_01_002: [The getDiagSamplingPercentage function shall return the value of sampling percentage.] */
+    return this.diagSamplingPercentage;
+  }
+
+  /**
+   * @method            module:azure-iot-device.DiagnosticClient.setDiagSamplingPercentage
+   * @description       set the diagnostic sampling percentage.
+   *
+   * @param {number}   diagSamplingPercentage   The value of sampling percentage.
+   * @throws {ArgumentError}     If value < 0 or value > 100.
+   * @throws {ArgumentError}     If value is not integer.
+   */
+  setDiagSamplingPercentage(diagSamplingPercentage: number): void {
+    /* Codes_SRS_NODE_DEVICE_DIAGNOSTICCLIENT_01_004: [The setDiagSamplingPercentage function shall throw an exception when input parameter is not an integer.] */
+    if (diagSamplingPercentage % 1 !== 0) {
+      throw new errors.ArgumentError('Sampling percentage should be integer');
```

Comment:
>throw [](start = 6, length = 5)

this will crash the whole device client. I understand if we were to ignore incorrect values, but crashing the client based on input coming from the service isn't acceptable especially if there is no validation on the service side